Compile failure when using generic type alias and providing int to function directly

### Describe the bug

Below code snippet fails if you pass an int directly `println(something(5))` but works if you pass in a variable with the same value assigned to it `println(something(new_var))`. It also complies if you pass Something{} or string to it directly or via variables, it only fails to compile when passing int value directly

```
/Users/brent/aaaa/v/v ../monolith -color -g -keepc
monolith.v:11:2: warning: unused variable: `new_var`
    9 | 
   10 | fn main() {
   11 |     new_var := 5
      |     ~~~~~~~
   12 |     println(something(5))
   13 | }
warning: tcc compilation failed, falling back to cc (this is much slower)
/private/tmp/v_501/monolith.tmp.c:1998: warning: type defaults to int
/private/tmp/v_501/monolith.tmp.c:1998: error: ';' expected (got 'int_to_sumtype_main__Any_T_int')
================== C compiler bug report ==============
Sent C compiler bug report to https://bugs.vlang.io/bug-report.
Report id: 66652a31-4391-4db7-94be-479740b9a85f
Delete this report from the server with:
  curl -X DELETE 'https://bugs.vlang.io/bug-report/66652a31-4391-4db7-94be-479740b9a85f?token=48b7603c-1c90-4a68-914c-43692fe9d0d2'
Generated C lines sent from /private/tmp/v_501/monolith.tmp.c:1998:
    1993 | void builtin__ArrayFlags_clear(ArrayFlags* e, ArrayFlags flag_);
    1994 | void main__main(void);
    1995 | int main__something_T_int(main__Any_T_intliteral my_var);
    1996 | static string main__Any_T_intliteral_str(main__Any_T_intliteral x);
    1997 | static string indent_main__Any_T_intliteral_str(main__Any_T_intliteral x, int indent_count);
>   1998 | main__Any_T_int int_to_sumtype_main__Any_T_int(int* x, bool is_mut);
    1999 | 
    2000 | // V global/const non-precomputed definitions:
    2001 | static string _const_strconv__digit_pairs; // a string literal, inited later
    2002 | static string _const_strconv__base_digits; // a string literal, inited later
    2003 | static string _const_grapheme_control_ranges; // a string literal, inited later
Corresponding V lines sent: no V source mapping was available.
=======================================================
builder error: 
==================
C error found while compiling generated C code.
This can be caused by invalid C interop code, C compiler flags, or a V compiler bug.
If your code is pure V and this still happens, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Reproduction Steps

Below will not compile, but if you insted pass new_var to the something func it will compile
```v
struct Something {}
type Any[T] = string | int | T

fn something[T](my_var Any[T]) int {
	println(my_var)
	return 0
}

fn main() {
	new_var := 5
	println(something(5))
}
```

### Expected Behavior

To compile for both ways

### Current Behavior

only compiles when variable is passed in 

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

Current V version: V 0.5.1 1704098, timestamp: 2026-06-21 02:28:18 +0300

### Environment details (OS name and version, etc.)

brent@Brents-MacBook-Pro monorepo % v doctor
|V full version      |V 0.5.1 c9c938645214280ea894abfffc184a0f893cffd6.1704098
|:-------------------|:-------------------
|OS                  |macos, macOS, 26.4.1, 25E253
|Processor           |12 cpus, 64bit, little endian, Apple M4 Pro
|Memory              |0.42GB/24GB
|                    |
|V executable        |/Users/brent/aaaa/v/v
|V last modified time|2026-06-21 00:25:28
|                    |
|V home dir          |OK, value: /Users/brent/aaaa/v
|VMODULES            |OK, value: /Users/brent/.vmodules
|VTMP                |OK, value: /tmp/v_501
|Current working dir |OK, value: /Users/brent/aaaa/monorepo
|                    |
|Git version         |git version 2.50.1 (Apple Git-155)
|V git status        |1704098d
|.git/config present |true
|                    |
|cc version          |Apple clang version 21.0.0 (clang-2100.0.123.102)
|gcc version         |Apple clang version 21.0.0 (clang-2100.0.123.102)
|clang version       |Apple clang version 21.0.0 (clang-2100.0.123.102)
|tcc version         |tcc version 0.9.28rc 2026-01-10 HEAD@5ec0e6f8 (AArch64 Darwin)
|tcc git status      |thirdparty-macos-arm64 f995efa3
|emcc version        |N/A
|glibc version       |N/A
brent@Brents-MacBook-Pro monorepo % 
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/builder/builder.v
+++ b/vlib/v/builder/builder.v
@@ -284,6 +284,7 @@ pub fn (mut b Builder) parse_imports() {
 		util.timing_measure(@METHOD)
 	}
 	mut done_imports := []string{}
+	mut done_import_paths := map[string]bool{}
 	if b.pref.is_vsh {
 		done_imports << 'os'
 	}
@@ -331,6 +332,11 @@ pub fn (mut b Builder) parse_imports() {
 					ast_file.path, imp.pos)
 				break
 			}
+			import_path_key := comparable_real_path(import_path)
+			if import_path_key in done_import_paths {
+				done_imports << mod
+				continue
+			}
 			v_files := b.v_files_from_dir(import_path)
 			if v_files.len == 0 {
 				// v.parsers[i].error_with_token_index('cannot import module "${mod}" (no .v files in "${import_path}")', v.parsers[i].import_ast.get_import_tok_idx(mod))
@@ -358,6 +364,7 @@ pub fn (mut b Builder) parse_imports() {
 				return
 			}
 			done_imports << mod
+			done_import_paths[import_path_key] = true
 		}
 	}
 	b.resolve_deps()

--- a/vlib/v/builder/builder.v
+++ b/vlib/v/builder/builder.v
@@ -284,7 +284,8 @@ pub fn (mut b Builder) parse_imports() {
 		util.timing_measure(@METHOD)
 	}
 	mut done_imports := []string{}
-	mut done_import_paths := map[string]bool{}
+	mut done_import_path_modules := map[string][]string{}
+	mut done_import_path_files := map[string][]string{}
 	if b.pref.is_vsh {
 		done_imports << 'os'
 	}
@@ -333,7 +334,11 @@ pub fn (mut b Builder) parse_imports() {
 				break
 			}
 			import_path_key := comparable_real_path(import_path)
-			if import_path_key in done_import_paths {
+			if import_path_key in done_import_path_modules {
+				for module_idx, name in done_import_path_modules[import_path_key] {
+					b.validate_imported_module_name(i, ast_file.path, mod,
+						done_import_path_files[import_path_key][module_idx], name, imp.pos)
+				}
 				done_imports << mod
 				continue
 			}
@@ -347,24 +352,21 @@ pub fn (mut b Builder) parse_imports() {
 			// eprintln('>> ast_file.path: ${ast_file.path} , done: ${done_imports}, `import ${mod}` => ${v_files}')
 			// Add all imports referenced by these libs
 			parsed_files := parser.parse_files(v_files, mut b.table, b.pref)
+			mut imported_module_names := []string{}
+			mut imported_module_paths := []string{}
 			for file in parsed_files {
-				mut name := file.mod.name
-				if name == '' {
-					name = file.mod.short_name
-				}
-				sname := name.all_after_last('.')
-				smod := mod.all_after_last('.')
-				if sname != smod {
-					msg := 'bad module definition: ${ast_file.path} imports module "${mod}" but ${file.path} is defined as module `${name}`'
-					b.parsed_files[i].errors << b.error_with_pos(msg, ast_file.path, imp.pos)
-				}
+				name := imported_file_module_name(file)
+				b.validate_imported_module_name(i, ast_file.path, mod, file.path, name, imp.pos)
+				imported_module_names << name
+				imported_module_paths << file.path
 			}
 			b.parsed_files << parsed_files
 			if b.should_stop_after_frontend_error() && parsed_files.any(it.errors.len > 0) {
 				return
 			}
 			done_imports << mod
-			done_import_paths[import_path_key] = true
+			done_import_path_modules[import_path_key] = imported_module_names
+			done_import_path_files[import_path_key] = imported_module_paths
 		}
 	}
 	b.resolve_deps()
@@ -435,6 +437,22 @@ pub fn (mut b Builder) resolve_deps() {
 		}
 		b.table.modules = mods
 		b.parsed_files = reordered_parsed_files
+	}
+}
+
+fn imported_file_module_name(file ast.File) string {
+	if file.mod.name != '' {
+		return file.mod.name
+	}
+	return file.mod.short_name
+}
+
+fn (mut b Builder) validate_imported_module_name(importer_idx int, importer_path string, mod string, imported_path string, name string, import_pos token.Pos) {
+	sname := name.all_after_last('.')
+	smod := mod.all_after_last('.')
+	if sname != smod {
+		msg := 'bad module definition: ${importer_path} imports module "${mod}" but ${imported_path} is defined as module `${name}`'
+		b.parsed_files[importer_idx].errors << b.error_with_pos(msg, importer_path, import_pos)
 	}
 }
 

--- a/vlib/v/builder/builder_test.v
+++ b/vlib/v/builder/builder_test.v
@@ -255,6 +255,47 @@ pub struct Client {}
 	assert !res.output.contains('cannot register struct')
 }
 
+fn test_duplicate_resolved_import_path_still_validates_module_name() {
+	$if windows {
+		return
+	}
+	os.chdir(test_path)!
+	workspace := os.join_path(test_path, 'run_duplicate_resolved_import_path_validation')
+	project_dir := os.join_path(workspace, 'project')
+	foo_dir := os.join_path(project_dir, 'modules', 'foo')
+	bar_dir := os.join_path(project_dir, 'modules', 'bar')
+	defer {
+		os.chdir(test_path) or {}
+		os.rmdir_all(workspace) or {}
+	}
+	os.mkdir_all(foo_dir)!
+	os.write_file(os.join_path(project_dir, 'v.mod'), "Module {\n\tname: 'project'\n}\n")!
+	os.write_file(os.join_path(foo_dir, 'foo.v'), 'module foo
+
+pub fn value() int {
+	return 1
+}
+')!
+	os.symlink(foo_dir, bar_dir) or { return }
+	os.write_file(os.join_path(project_dir, 'main.v'), 'module main
+
+import foo
+import bar
+
+fn main() {
+	println(foo.value())
+	println(bar.value())
+}
+')!
+	os.chdir(project_dir)!
+
+	res := os.execute('${os.quoted_path(vexe)} -check main.v')
+	assert res.exit_code != 0, res.output
+	assert res.output.contains('bad module definition'), res.output
+	assert res.output.contains('imports module "bar"'), res.output
+	assert res.output.contains('defined as module `foo`'), res.output
+}
+
 fn test_removed_src_layout_error_mentions_vmod_subdirs() {
 	os.chdir(test_path)!
 	project_dir := os.join_path(test_path, 'run_removed_src_project')

--- a/vlib/v/builder/builder_test.v
+++ b/vlib/v/builder/builder_test.v
@@ -216,6 +216,45 @@ fn test_empty_local_dir_does_not_shadow_vlib_module() {
 	assert res.output.trim_space() == 'true'
 }
 
+fn test_imports_with_same_resolved_path_are_parsed_once() {
+	os.chdir(test_path)!
+	workspace := os.join_path(test_path, 'run_duplicate_resolved_import_path')
+	project_dir := os.join_path(workspace, 'dupmod')
+	defer {
+		os.chdir(test_path) or {}
+		os.rmdir_all(workspace) or {}
+	}
+	os.mkdir_all(os.join_path(project_dir, 'providers'))!
+	os.write_file(os.join_path(project_dir, 'v.mod'), "Module {\n\tname: 'dupmod'\n}\n")!
+	os.write_file(os.join_path(project_dir, 'dupmod.v'), 'module dupmod
+
+import providers
+
+pub fn make_client() providers.Client {
+	return providers.Client{}
+}
+')!
+	os.write_file(os.join_path(project_dir, 'main_test.v'), 'module main
+
+import dupmod
+import dupmod.providers
+
+fn test_client() {
+	_ := dupmod.make_client()
+	_ := providers.Client{}
+}
+')!
+	os.write_file(os.join_path(project_dir, 'providers', 'client.v'), 'module providers
+
+pub struct Client {}
+')!
+	os.chdir(project_dir)!
+
+	res := os.execute('${os.quoted_path(vexe)} -check main_test.v')
+	assert res.exit_code == 0, res.output
+	assert !res.output.contains('cannot register struct')
+}
+
 fn test_removed_src_layout_error_mentions_vmod_subdirs() {
 	os.chdir(test_path)!
 	project_dir := os.join_path(test_path, 'run_removed_src_project')

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -1908,8 +1908,9 @@ fn (mut c Checker) infer_fn_generic_types(func &ast.Fn, mut node ast.CallExpr) {
 			s := c.table.type_to_str(typ)
 			println('inferred `${func.name}[${s}]`')
 		}
-		inferred_types << c.unwrap_generic(typ)
-		node.concrete_types << typ
+		concrete_typ := ast.mktyp(c.unwrap_generic(typ))
+		inferred_types << concrete_typ
+		node.concrete_types << concrete_typ
 	}
 
 	if c.table.register_fn_concrete_types(func.fkey(), inferred_types) {

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -763,19 +763,29 @@ fn (mut c Checker) match_sumtype_variant_is_handled_by(variant ast.Type, handled
 }
 
 fn (mut c Checker) match_sumtype_variant_is_handled(variant ast.Type, handled ast.Type) bool {
-	unaliased_variant := c.table.fully_unaliased_type(variant)
-	unaliased_handled := c.table.fully_unaliased_type(handled)
-	variant_sym := c.table.sym(unaliased_variant)
-	handled_sym := c.table.sym(unaliased_handled)
+	if variant.idx() == handled.idx() && variant.has_flag(.option) == handled.has_flag(.option)
+		&& variant.nr_muls() == handled.nr_muls() {
+		return true
+	}
+	handled_sym := c.table.sym(handled)
+	if handled_sym.info is ast.Alias {
+		mut parent_type := handled_sym.info.parent_type.set_nr_muls(handled.nr_muls())
+		if handled.has_flag(.option) {
+			parent_type = parent_type.set_flag(.option)
+		}
+		if handled.has_flag(.result) {
+			parent_type = parent_type.set_flag(.result)
+		}
+		return c.match_sumtype_variant_is_handled(variant, parent_type)
+	}
+	variant_sym := c.table.sym(variant)
 	if variant_sym.info is ast.FnType && handled_sym.info is ast.FnType {
 		return
 			c.table.fn_type_source_signature(variant_sym.info.func) == c.table.fn_type_source_signature(handled_sym.info.func)
-			&& unaliased_variant.has_flag(.option) == unaliased_handled.has_flag(.option)
-			&& unaliased_variant.nr_muls() == unaliased_handled.nr_muls()
+			&& variant.has_flag(.option) == handled.has_flag(.option)
+			&& variant.nr_muls() == handled.nr_muls()
 	}
-	return unaliased_variant.idx() == unaliased_handled.idx()
-		&& unaliased_variant.has_flag(.option) == unaliased_handled.has_flag(.option)
-		&& unaliased_variant.nr_muls() == unaliased_handled.nr_muls()
+	return false
 }
 
 fn (mut c Checker) match_exprs(mut node ast.MatchExpr, cond_type_sym ast.TypeSymbol, cond_final_sym ast.TypeSymbol) {

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -684,6 +684,49 @@ fn (mut c Checker) get_match_case_literal_value(mut expr ast.Expr) ?i64 {
 	return none
 }
 
+fn (mut c Checker) match_sumtype_has_variant(parent ast.Type, variant ast.Type) bool {
+	if c.table.sumtype_has_variant_recursive(parent, variant, true) {
+		return true
+	}
+	if c.table.sym(parent).kind == .sum_type {
+		return false
+	}
+	for candidate in c.concrete_sumtype_variants(parent) {
+		if c.match_sumtype_variant_is_handled(candidate, variant) {
+			return true
+		}
+	}
+	return false
+}
+
+fn (mut c Checker) match_sumtype_missing_variants(parent ast.Type, handled []ast.Type) []ast.Type {
+	if c.table.sym(parent).kind == .sum_type {
+		return c.table.sumtype_missing_variants(parent, handled)
+	}
+	mut missing := []ast.Type{}
+	for variant in c.concrete_sumtype_variants(parent) {
+		mut is_handled := false
+		for handled_variant in handled {
+			if c.match_sumtype_variant_is_handled(variant, handled_variant) {
+				is_handled = true
+				break
+			}
+		}
+		if !is_handled {
+			missing << variant
+		}
+	}
+	return missing
+}
+
+fn (mut c Checker) match_sumtype_variant_is_handled(variant ast.Type, handled ast.Type) bool {
+	unaliased_variant := c.table.fully_unaliased_type(variant)
+	unaliased_handled := c.table.fully_unaliased_type(handled)
+	return unaliased_variant.idx() == unaliased_handled.idx()
+		&& unaliased_variant.has_flag(.option) == unaliased_handled.has_flag(.option)
+		&& unaliased_variant.nr_muls() == unaliased_handled.nr_muls()
+}
+
 fn (mut c Checker) match_exprs(mut node ast.MatchExpr, cond_type_sym ast.TypeSymbol, cond_final_sym ast.TypeSymbol) {
 	c.expected_type = node.expected_type
 	if node.cond_type.idx() == 0 {
@@ -694,6 +737,8 @@ fn (mut c Checker) match_exprs(mut node ast.MatchExpr, cond_type_sym ast.TypeSym
 	is_alias_to_matchable_type := cond_type_sym.kind == .alias
 		&& cond_final_sym.kind in [.interface, .sum_type]
 	cond_match_sym := if is_alias_to_matchable_type { cond_final_sym } else { cond_type_sym }
+	sumtype_match_variants := c.concrete_sumtype_variants(cond_match_type)
+	is_cond_match_sumtype := sumtype_match_variants.len > 0
 	mut enum_ref_checked := false
 	mut is_comptime_value_match := false
 	// branch_exprs is a histogram of how many times
@@ -908,12 +953,12 @@ fn (mut c Checker) match_exprs(mut node ast.MatchExpr, cond_type_sym ast.TypeSym
 							}
 						}
 					}
-				} else if cond_match_sym.info is ast.SumType {
-					if !c.table.sumtype_has_variant_recursive(cond_match_type, expr_type, true) {
+				} else if is_cond_match_sumtype {
+					if !c.match_sumtype_has_variant(cond_match_type, expr_type) {
 						expr_str := c.table.type_to_str(expr_type)
 						expect_str := c.table.type_to_str(node.cond_type)
 						sumtype_variant_names :=
-							c.table.sumtype_matchable_variants(cond_match_type).map(c.table.type_to_str_using_aliases(it, {}))
+							sumtype_match_variants.map(c.table.type_to_str_using_aliases(it, {}))
 						suggestion := util.new_suggestion(expr_str, sumtype_variant_names)
 						c.error(suggestion.say('`${expect_str}` has no variant `${expr_str}`'),
 							expr.pos())
@@ -941,7 +986,7 @@ fn (mut c Checker) match_exprs(mut node ast.MatchExpr, cond_type_sym ast.TypeSym
 		}
 		// when match is type matching, then register smart cast for every branch
 		if expr_types.len > 0 {
-			if cond_match_sym.kind in [.sum_type, .interface] {
+			if is_cond_match_sumtype || cond_match_sym.kind == .interface {
 				mut expr_type := ast.no_type
 				if expr_types.len > 1 {
 					mut agg_name := strings.new_builder(20)
@@ -999,33 +1044,33 @@ fn (mut c Checker) match_exprs(mut node ast.MatchExpr, cond_type_sym ast.TypeSym
 			}
 		}
 	} else {
-		match cond_match_sym.info {
-			ast.SumType {
-				for v in c.table.sumtype_missing_variants(cond_match_type, branch_expr_types) {
-					is_exhaustive = false
-					unhandled << '`${c.table.type_to_str(v)}`'
-				}
+		if is_cond_match_sumtype {
+			for v in c.match_sumtype_missing_variants(cond_match_type, branch_expr_types) {
+				is_exhaustive = false
+				unhandled << '`${c.table.type_to_str(v)}`'
 			}
-			//
-			ast.Enum {
-				for v in cond_match_sym.info.vals {
-					mut is_handled := v in branch_exprs
-					if !is_handled && is_multi_allowed_enum_match {
-						if enum_val := c.table.find_enum_field_val(cond_match_sym.name, v) {
-							is_handled = enum_val in branch_enum_values
+		} else {
+			match cond_match_sym.info {
+				ast.Enum {
+					for v in cond_match_sym.info.vals {
+						mut is_handled := v in branch_exprs
+						if !is_handled && is_multi_allowed_enum_match {
+							if enum_val := c.table.find_enum_field_val(cond_match_sym.name, v) {
+								is_handled = enum_val in branch_enum_values
+							}
+						}
+						if !is_handled {
+							is_exhaustive = false
+							unhandled << '`.${v}`'
 						}
 					}
-					if !is_handled {
+					if cond_match_sym.info.is_flag {
 						is_exhaustive = false
-						unhandled << '`.${v}`'
 					}
 				}
-				if cond_match_sym.info.is_flag {
+				else {
 					is_exhaustive = false
 				}
-			}
-			else {
-				is_exhaustive = false
 			}
 		}
 	}

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -765,6 +765,14 @@ fn (mut c Checker) match_sumtype_variant_is_handled_by(variant ast.Type, handled
 fn (mut c Checker) match_sumtype_variant_is_handled(variant ast.Type, handled ast.Type) bool {
 	unaliased_variant := c.table.fully_unaliased_type(variant)
 	unaliased_handled := c.table.fully_unaliased_type(handled)
+	variant_sym := c.table.sym(unaliased_variant)
+	handled_sym := c.table.sym(unaliased_handled)
+	if variant_sym.info is ast.FnType && handled_sym.info is ast.FnType {
+		return
+			c.table.fn_type_source_signature(variant_sym.info.func) == c.table.fn_type_source_signature(handled_sym.info.func)
+			&& unaliased_variant.has_flag(.option) == unaliased_handled.has_flag(.option)
+			&& unaliased_variant.nr_muls() == unaliased_handled.nr_muls()
+	}
 	return unaliased_variant.idx() == unaliased_handled.idx()
 		&& unaliased_variant.has_flag(.option) == unaliased_handled.has_flag(.option)
 		&& unaliased_variant.nr_muls() == unaliased_handled.nr_muls()

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -691,7 +691,7 @@ fn (mut c Checker) match_sumtype_has_variant(parent ast.Type, variant ast.Type) 
 	if c.table.sym(parent).kind == .sum_type {
 		return false
 	}
-	for candidate in c.concrete_sumtype_variants(parent) {
+	for candidate in c.match_sumtype_matchable_variants(parent) {
 		if c.match_sumtype_variant_is_handled(candidate, variant) {
 			return true
 		}
@@ -699,24 +699,67 @@ fn (mut c Checker) match_sumtype_has_variant(parent ast.Type, variant ast.Type) 
 	return false
 }
 
+fn (mut c Checker) match_sumtype_matchable_variants(parent ast.Type) []ast.Type {
+	if c.table.sym(parent).kind == .sum_type {
+		return c.table.sumtype_matchable_variants(parent)
+	}
+	mut variants := []ast.Type{}
+	mut seen := map[u32]bool{}
+	c.collect_match_sumtype_matchable_variants(parent, mut seen, mut variants)
+	return variants
+}
+
+fn (mut c Checker) collect_match_sumtype_matchable_variants(parent ast.Type, mut seen map[u32]bool, mut variants []ast.Type) {
+	for variant in c.concrete_sumtype_variants(parent) {
+		key := u32(variant)
+		if key in seen {
+			continue
+		}
+		seen[key] = true
+		variants << variant
+		if c.concrete_sumtype_variants(variant).len > 0 {
+			c.collect_match_sumtype_matchable_variants(variant, mut seen, mut variants)
+		}
+	}
+}
+
 fn (mut c Checker) match_sumtype_missing_variants(parent ast.Type, handled []ast.Type) []ast.Type {
 	if c.table.sym(parent).kind == .sum_type {
 		return c.table.sumtype_missing_variants(parent, handled)
 	}
 	mut missing := []ast.Type{}
+	mut seen := map[u32]bool{}
+	c.collect_match_sumtype_missing_variants(parent, handled, mut seen, mut missing)
+	return missing
+}
+
+fn (mut c Checker) collect_match_sumtype_missing_variants(parent ast.Type, handled []ast.Type, mut seen map[u32]bool, mut missing []ast.Type) {
+	if c.match_sumtype_variant_is_handled_by(parent, handled) {
+		return
+	}
 	for variant in c.concrete_sumtype_variants(parent) {
-		mut is_handled := false
-		for handled_variant in handled {
-			if c.match_sumtype_variant_is_handled(variant, handled_variant) {
-				is_handled = true
-				break
+		if c.match_sumtype_variant_is_handled_by(variant, handled) {
+			continue
+		}
+		if c.concrete_sumtype_variants(variant).len > 0 {
+			c.collect_match_sumtype_missing_variants(variant, handled, mut seen, mut missing)
+		} else {
+			key := u32(variant)
+			if key !in seen {
+				seen[key] = true
+				missing << variant
 			}
 		}
-		if !is_handled {
-			missing << variant
+	}
+}
+
+fn (mut c Checker) match_sumtype_variant_is_handled_by(variant ast.Type, handled []ast.Type) bool {
+	for handled_variant in handled {
+		if c.match_sumtype_variant_is_handled(variant, handled_variant) {
+			return true
 		}
 	}
-	return missing
+	return false
 }
 
 fn (mut c Checker) match_sumtype_variant_is_handled(variant ast.Type, handled ast.Type) bool {
@@ -737,7 +780,7 @@ fn (mut c Checker) match_exprs(mut node ast.MatchExpr, cond_type_sym ast.TypeSym
 	is_alias_to_matchable_type := cond_type_sym.kind == .alias
 		&& cond_final_sym.kind in [.interface, .sum_type]
 	cond_match_sym := if is_alias_to_matchable_type { cond_final_sym } else { cond_type_sym }
-	sumtype_match_variants := c.concrete_sumtype_variants(cond_match_type)
+	sumtype_match_variants := c.match_sumtype_matchable_variants(cond_match_type)
 	is_cond_match_sumtype := sumtype_match_variants.len > 0
 	mut enum_ref_checked := false
 	mut is_comptime_value_match := false

--- a/vlib/v/checker/tests/generic_sumtype_alias_distinct_variant_match.out
+++ b/vlib/v/checker/tests/generic_sumtype_alias_distinct_variant_match.out
@@ -1,0 +1,15 @@
+vlib/v/checker/tests/generic_sumtype_alias_distinct_variant_match.vv:8:3: error: `GenericAliasBox[string]` has no variant `GenericAliasB`.
+Did you mean `GenericAliasA`?
+    6 |     value := GenericAliasBox[string](GenericAliasA(1))
+    7 |     match value {
+    8 |         GenericAliasB {
+      |         ~~~~~~~~~~~~~
+    9 |             println('bad')
+   10 |         }
+vlib/v/checker/tests/generic_sumtype_alias_distinct_variant_match.vv:7:2: error: match must be exhaustive (add match branches for: `GenericAliasA` or `else {}` at the end)
+    5 | fn main() {
+    6 |     value := GenericAliasBox[string](GenericAliasA(1))
+    7 |     match value {
+      |     ~~~~~~~~~~~~~
+    8 |         GenericAliasB {
+    9 |             println('bad')

--- a/vlib/v/checker/tests/generic_sumtype_alias_distinct_variant_match.vv
+++ b/vlib/v/checker/tests/generic_sumtype_alias_distinct_variant_match.vv
@@ -1,0 +1,15 @@
+type GenericAliasA = int
+type GenericAliasB = int
+type GenericAliasBox[T] = GenericAliasA | T
+
+fn main() {
+	value := GenericAliasBox[string](GenericAliasA(1))
+	match value {
+		GenericAliasB {
+			println('bad')
+		}
+		string {
+			println(value)
+		}
+	}
+}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -410,8 +410,9 @@ fn (mut g Gen) resolve_current_fn_generic_type(typ ast.Type) ast.Type {
 	if generic_names.len == 0 || g.cur_concrete_types.len == 0 {
 		return g.unwrap_generic(typ)
 	}
+	concrete_types := g.cur_concrete_types.map(ast.mktyp(it))
 	mut muttable := unsafe { &ast.Table(g.table) }
-	if resolved := muttable.convert_generic_type(typ, generic_names, g.cur_concrete_types) {
+	if resolved := muttable.convert_generic_type(typ, generic_names, concrete_types) {
 		return g.unwrap_generic(resolved)
 	}
 	return g.unwrap_generic(g.recheck_concrete_type(typ))
@@ -2017,7 +2018,13 @@ fn (mut g Gen) fn_decl_params(params []ast.Param, scope &ast.Scope, is_variadic 
 		if i >= param_count {
 			break
 		}
-		mut typ := g.unwrap_generic(param.typ)
+		mut typ := if !param.typ.has_flag(.variadic) && g.cur_concrete_types.len > 0
+			&& param.orig_typ != 0 && (param.orig_typ.has_flag(.generic)
+			|| g.type_has_unresolved_generic_parts(param.orig_typ)) {
+			g.resolve_current_fn_generic_type(param.orig_typ)
+		} else {
+			g.unwrap_generic(param.typ)
+		}
 		if g.pref.translated && g.file.is_translated && param.typ.has_flag(.variadic) {
 			typ = g.table.sym(typ).array_info().elem_type.set_flag(.variadic)
 		}

--- a/vlib/v/tests/sumtypes/generic_sumtype_alias_match_regression_test.v
+++ b/vlib/v/tests/sumtypes/generic_sumtype_alias_match_regression_test.v
@@ -1,0 +1,15 @@
+struct GenericSumtypeAliasMatchItem {}
+
+type GenericSumtypeAliasMatchValue[T] = string | int | T
+
+fn generic_sumtype_alias_match_exhaustive[T](value GenericSumtypeAliasMatchValue[T]) int {
+	return match value {
+		string { 1 }
+		int { 2 }
+		T { 3 }
+	}
+}
+
+fn test_generic_sumtype_alias_match_with_generic_variant_is_exhaustive() {
+	assert generic_sumtype_alias_match_exhaustive(GenericSumtypeAliasMatchItem{}) == 3
+}

--- a/vlib/v/tests/sumtypes/generic_sumtype_alias_match_regression_test.v
+++ b/vlib/v/tests/sumtypes/generic_sumtype_alias_match_regression_test.v
@@ -29,3 +29,20 @@ fn test_generic_sumtype_alias_match_expands_nested_sumtype_variants() {
 	assert generic_sumtype_alias_match_nested_variants[bool](1) == 1
 	assert generic_sumtype_alias_match_nested_variants[bool](true) == 3
 }
+
+type GenericSumtypeAliasFnBox[T] = fn (int) int | T
+
+fn generic_sumtype_alias_match_increment(value int) int {
+	return value + 1
+}
+
+fn generic_sumtype_alias_match_fn_variant[T](value GenericSumtypeAliasFnBox[T]) int {
+	return match value {
+		fn (int) int { value(41) }
+		T { 0 }
+	}
+}
+
+fn test_generic_sumtype_alias_match_function_variant_by_signature() {
+	assert generic_sumtype_alias_match_fn_variant[string](generic_sumtype_alias_match_increment) == 42
+}

--- a/vlib/v/tests/sumtypes/generic_sumtype_alias_match_regression_test.v
+++ b/vlib/v/tests/sumtypes/generic_sumtype_alias_match_regression_test.v
@@ -13,3 +13,19 @@ fn generic_sumtype_alias_match_exhaustive[T](value GenericSumtypeAliasMatchValue
 fn test_generic_sumtype_alias_match_with_generic_variant_is_exhaustive() {
 	assert generic_sumtype_alias_match_exhaustive(GenericSumtypeAliasMatchItem{}) == 3
 }
+
+type GenericSumtypeAliasNestedInner = int | string
+type GenericSumtypeAliasNestedOuter[T] = GenericSumtypeAliasNestedInner | T
+
+fn generic_sumtype_alias_match_nested_variants[T](value GenericSumtypeAliasNestedOuter[T]) int {
+	return match value {
+		int { 1 }
+		string { 2 }
+		T { 3 }
+	}
+}
+
+fn test_generic_sumtype_alias_match_expands_nested_sumtype_variants() {
+	assert generic_sumtype_alias_match_nested_variants[bool](1) == 1
+	assert generic_sumtype_alias_match_nested_variants[bool](true) == 3
+}

--- a/vlib/v/tests/sumtypes/generic_sumtype_alias_regression_test.v
+++ b/vlib/v/tests/sumtypes/generic_sumtype_alias_regression_test.v
@@ -1,0 +1,10 @@
+type GenericSumtypeAliasValue[T] = string | int | T
+
+fn generic_sumtype_alias_accepts_int_literal[T](value GenericSumtypeAliasValue[T]) int {
+	_ = value
+	return 0
+}
+
+fn test_generic_sumtype_alias_accepts_int_literal() {
+	assert generic_sumtype_alias_accepts_int_literal(5) == 0
+}

--- a/vlib/v/util/module.v
+++ b/vlib/v/util/module.v
@@ -28,7 +28,7 @@ pub fn qualify_import(pref_ &pref.Preferences, mod string, file_path string) str
 	for search_path in mod_paths {
 		try_path := os.join_path_single(search_path, mod_path)
 		if os.is_dir(try_path) {
-			if m1 := mod_path_to_full_name(pref_, mod, try_path) {
+			if m1 := import_path_to_full_name(pref_, mod, try_path) {
 				trace_qualify(@FN, mod, file_path, 'import_res 1', m1, try_path)
 				// >  qualify_import: term | file_path: /v/vls/server/diagnostics.v | => import_res 1: term  ; /v/cleanv/vlib/term
 				return m1
@@ -41,7 +41,7 @@ pub fn qualify_import(pref_ &pref.Preferences, mod string, file_path string) str
 	} else {
 		os.join_path_single(os.getwd(), file_path)
 	}
-	if m1 := mod_path_to_full_name(pref_, mod, abs_file_path) {
+	if m1 := import_path_to_full_name(pref_, mod, abs_file_path) {
 		trace_qualify(@FN, mod, file_path, 'import_res 2', m1, abs_file_path)
 		// >  qualify_module: analyzer           | file_path: /v/vls/analyzer/store.v  | =>   module_res 2: analyzer           ; clean_file_path - getwd == mod
 		// >  qualify_import: analyzer.depgraph  | file_path: /v/vls/analyzer/store.v  | =>   import_res 2: analyzer.depgraph  ; /v/vls/analyzer/store.v
@@ -116,6 +116,14 @@ pub fn qualify_module(pref_ &pref.Preferences, mod string, file_path string) str
 // 2022-01-30 it leads to path differences, and the / version on windows triggers a module lookip bug,
 // 2022-01-30 leading to completely different errors)
 fn mod_path_to_full_name(pref_ &pref.Preferences, mod string, path string) !string {
+	return mod_path_to_full_name_with_options(pref_, mod, path, false)
+}
+
+fn import_path_to_full_name(pref_ &pref.Preferences, mod string, path string) !string {
+	return mod_path_to_full_name_with_options(pref_, mod, path, true)
+}
+
+fn mod_path_to_full_name_with_options(pref_ &pref.Preferences, mod string, path string, allow_shorter_name bool) !string {
 	// TODO: explore using `pref.lookup_path` & `os.vmodules_paths()`
 	// absolute paths instead of 'vlib' & '.vmodules'
 	mut vmod_folders := ['vlib', '.vmodules', 'modules']
@@ -169,7 +177,7 @@ fn mod_path_to_full_name(pref_ &pref.Preferences, mod string, path string) !stri
 						relative_parts := real_try_path.all_after(prefix).split(os.path_separator)
 						mod_full_name := normalize_base_url_mod_name(relative_parts.join('.'),
 							try_path)
-						if mod_full_name.len < mod.len {
+						if !allow_shorter_name && mod_full_name.len < mod.len {
 							return mod
 						}
 						if !module_name_has_empty_part(mod_full_name) {
@@ -197,7 +205,7 @@ fn mod_path_to_full_name(pref_ &pref.Preferences, mod string, path string) !stri
 				if last_v_mod > -1 {
 					mod_full_name := normalize_base_url_mod_name(try_path_parts[last_v_mod..].join('.'),
 						try_path)
-					if mod_full_name.len < mod.len {
+					if !allow_shorter_name && mod_full_name.len < mod.len {
 						return mod
 					}
 					if !module_name_has_empty_part(mod_full_name) {

